File: verification/scripts/r/diagnostics/test_durbin_watson.R
# ============================================================================
# Durbin-Watson Test Reference Implementation (R)
# ============================================================================
# This script generates reference values for the Durbin-Watson test using R's
# lmtest::dwtest function. The Durbin-Watson test checks for autocorrelation
# in the residuals.
#
# Source: lmtest package, dwtest function
# Reference: Durbin & Watson (1950), "Testing for Serial Correlation in
#            Least Squares Regression: I"
#            Biometrika, Vol. 37, pp. 409-428
#            (1951), "Testing for Serial Correlation in Least Squares
#            Regression: II", Biometrika, Vol. 38, pp. 159-178
#
# Usage:
#   Rscript test_durbin_watson.R [csv_path] [output_dir]
#   Args:
#     csv_path  - Path to CSV file (first col = response, rest = predictors)
#                 Default: ../../datasets/csv/mtcars.csv
#     output_dir- Path to output directory
#                 Default: ../../results/r
# ============================================================================

library(lmtest)
library(jsonlite)

# Helper function to convert categorical columns to numeric
convert_categorical_to_numeric <- function(data, dataset_name) {
  # Check for non-numeric columns
  non_numeric_cols <- names(data)[sapply(data, function(x) !is.numeric(x))]

  if (length(non_numeric_cols) > 0) {
    cat(paste0("INFO: Dataset '", dataset_name, "' contains non-numeric columns: ",
               paste(non_numeric_cols, collapse = ", "), "\n"))
    cat("Converting categorical variables to numeric representations...\n")

    for (col in non_numeric_cols) {
      if (is.factor(data[[col]])) {
        # For factors, use integer level encoding
        unique_vals <- length(unique(data[[col]]))
        data[[col]] <- as.numeric(data[[col]])
        cat(paste0("  ", col, ": ", unique_vals, " unique values -> integer level encoding\n"))
      } else if (is.character(data[[col]])) {
        # For character columns, first try to convert to numeric directly
        # If that produces NAs, convert to factor and use integer encoding
        unique_vals <- length(unique(data[[col]]))
        temp_numeric <- as.numeric(data[[col]])
        if (any(is.na(temp_numeric))) {
          # Contains non-numeric strings, use factor encoding
          data[[col]] <- as.numeric(as.factor(data[[col]]))
          cat(paste0("  ", col, ": ", unique_vals, " unique values -> integer level encoding\n"))
        } else {
          # Successfully converted to numeric
          data[[col]] <- temp_numeric
          cat(paste0("  ", col, ": ", unique_vals, " unique values -> numeric encoding\n"))
        }
      }
    }
  }

  return(data)
}

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Set defaults
default_csv <- "../../datasets/csv/mtcars.csv"
default_output <- "../../results/r"

# Parse arguments
csv_path <- ifelse(length(args) >= 1, args[1], default_csv)
output_dir <- ifelse(length(args) >= 2, args[2], default_output)

# Validate CSV path
if (!file.exists(csv_path)) {
  stop(paste("CSV file not found:", csv_path))
}

# Extract dataset name from filename
dataset_name <- tools::file_path_sans_ext(basename(csv_path))

# Read CSV data
data <- read.csv(csv_path)

# Convert categorical columns to numeric
data <- convert_categorical_to_numeric(data, dataset_name)

# Assume first column is response variable, rest are predictors
response_col <- names(data)[1]
predictor_cols <- names(data)[-1]

# Build formula: response ~ predictor1 + predictor2 + ...
formula_str <- paste(response_col, "~", paste(predictor_cols, collapse = " + "))
formula <- as.formula(formula_str)

# Fit the model
model <- lm(formula, data = data)

# Run Durbin-Watson test
dw_result <- dwtest(model)

# Determine interpretation
if (dw_result$statistic > 2.5) {
  interpretation <- "Possible negative autocorrelation"
} else if (dw_result$statistic < 1.5) {
  interpretation <- "Possible positive autocorrelation"
} else {
  interpretation <- "No significant autocorrelation"
}

# Print results
cat("Durbin-Watson Test (R - lmtest::dwtest)\n")
cat("======================================\n")
cat("Dataset:", dataset_name, "\n")
cat("Formula:", formula_str, "\n")
cat("DW-statistic:", dw_result$statistic, "\n")
cat("p-value:", dw_result$p.value, "\n")
cat("Interpretation:", interpretation, "\n")
cat("Passed:", dw_result$p.value > 0.05, "\n\n")

# Prepare output
output <- list(
  test_name = "Durbin-Watson Test (R - lmtest::dwtest)",
  dataset = dataset_name,
  formula = formula_str,
  statistic = as.numeric(dw_result$statistic),
  p_value = as.numeric(dw_result$p.value),
  passed = dw_result$p.value > 0.05,
  interpretation = interpretation,
  description = "Tests for autocorrelation in residuals. Values near 2 indicate no autocorrelation, values near 0 suggest positive autocorrelation, and values near 4 suggest negative autocorrelation. Uses lmtest::dwtest."
)

# Create output directory if it doesn't exist
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# Save to JSON with naming convention: {dataset}_durbin_watson.json
output_file <- file.path(output_dir, paste0(dataset_name, "_durbin_watson.json"))
write_json(output, output_file, pretty = TRUE, digits = 22)

cat("Results saved to:", output_file, "\n")

File: verification/scripts/python/diagnostics/test_durbin_watson.py
# ============================================================================
# Durbin-Watson Test Reference Implementation (Python)
# ============================================================================
# This script generates reference values for the Durbin-Watson test using
# statsmodels.stats.stattools.durbin_watson. The Durbin-Watson test checks
# for autocorrelation in the residuals.
#
# Source: statsmodels package, durbin_watson function
# Reference: Durbin & Watson (1950), "Testing for Serial Correlation in
#            Least Squares Regression: I", Biometrika, Vol. 37, pp. 409-428
#            (1951), "Testing for Serial Correlation in Least Squares
#            Regression: II", Biometrika, Vol. 38, pp. 159-178
#
# Usage:
#   python test_durbin_watson.py [--csv CSV_PATH] [--output-dir OUTPUT_DIR]
#
#   Args:
#     --csv       Path to CSV file (first col = response, rest = predictors)
#                 Default: ../../datasets/csv/mtcars.csv
#     --output-dir Path to output directory
#                 Default: ../../results/python
# ============================================================================

import argparse
import os
import json
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.stats.stattools import durbin_watson


def convert_categorical_to_numeric(data, dataset_name):
    """Convert categorical variables to numeric representations."""
    non_numeric_cols = data.select_dtypes(exclude=[np.number]).columns.tolist()

    if not non_numeric_cols:
        return []

    print(f"INFO: Dataset '{dataset_name}' contains non-numeric columns: {non_numeric_cols}")
    print(f"Converting categorical variables to numeric representations...")

    # Process each non-numeric column
    for col in non_numeric_cols:
        # Use factorize for reliable categorical encoding
        data[col], uniques = pd.factorize(data[col])
        print(f"  {col}: {len(uniques)} unique values -> integer level encoding")

    return non_numeric_cols


def main():
    parser = argparse.ArgumentParser(
        description="Durbin-Watson Test - Check autocorrelation using statsmodels"
    )
    parser.add_argument(
        "--csv",
        default="../../datasets/csv/mtcars.csv",
        help="Path to CSV file (first column = response, rest = predictors)"
    )
    parser.add_argument(
        "--output-dir",
        default="../../results/python",
        help="Path to output directory"
    )

    args = parser.parse_args()

    # Validate CSV path
    if not os.path.exists(args.csv):
        raise FileNotFoundError(f"CSV file not found: {args.csv}")

    # Extract dataset name from filename
    dataset_name = os.path.splitext(os.path.basename(args.csv))[0]

    # Read CSV data
    data = pd.read_csv(args.csv)

    # If there are non-numeric columns, convert them to numeric
    non_numeric_cols = data.select_dtypes(exclude=[np.number]).columns.tolist()
    if non_numeric_cols:
        convert_categorical_to_numeric(data, dataset_name)

    # After conversion, re-check if we still have non-numeric columns
    remaining_non_numeric = data.select_dtypes(exclude=[np.number]).columns.tolist()
    if remaining_non_numeric:
        # These could be things like empty strings or other non-numeric types that couldn't be converted
        raise ValueError(f"Could not convert the following non-numeric columns to numeric: {remaining_non_numeric}")

    # Assume first column is response variable, rest are predictors
    response_col = data.columns[0]
    predictor_cols = data.columns[1:]

    # Prepare data for statsmodels (add constant for intercept)
    X = data[predictor_cols]
    X = sm.add_constant(X)
    y = data[response_col]

    # Build formula string for output
    formula_str = f"{response_col} ~ {' + '.join(predictor_cols)}"

    # Fit the model
    model = sm.OLS(y, X).fit()

    # Run Durbin-Watson test
    # Returns DW statistic (no p-value from statsmodels)
    dw_stat = durbin_watson(model.resid)

    # Interpret result (values near 2 = no autocorrelation)
    if dw_stat > 2:
        interpretation = "No positive autocorrelation"
    elif dw_stat < 2:
        interpretation = "Possible positive autocorrelation"
    else:
        interpretation = "Inconclusive"

    # Print results
    print("Durbin-Watson Test (Python - statsmodels)")
    print("=" * 42)
    print(f"Dataset: {dataset_name}")
    print(f"Formula: {formula_str}")
    print(f"DW-statistic: {dw_stat}")
    print(f"Interpretation: {interpretation}")
    print(f"Passed (1.5 < d < 2.5): {1.5 < dw_stat < 2.5}")
    print()

    # Prepare output
    output = {
        "test_name": "Durbin-Watson Test (Python - statsmodels)",
        "dataset": dataset_name,
        "formula": formula_str,
        "statistic": float(dw_stat),
        "p_value": None,  # statsmodels doesn't provide p-value for DW test
        "passed": bool(1.5 < dw_stat < 2.5),
        "interpretation": interpretation,
        "description": "Tests for autocorrelation in residuals. Values near 2 indicate no autocorrelation, values near 0 suggest positive autocorrelation, and values near 4 suggest negative autocorrelation. Uses statsmodels.stats.stattools.durbin_watson."
    }

    # Create output directory if it doesn't exist
    os.makedirs(args.output_dir, exist_ok=True)

    # Save to JSON with naming convention: {dataset}_durbin_watson.json
    output_file = os.path.join(args.output_dir, f"{dataset_name}_durbin_watson.json")

    with open(output_file, 'w') as f:
        json.dump(output, f, indent=2)

    print(f"Results saved to: {output_file}")


if __name__ == "__main__":
    main()
